src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineControl_ModelComputation.cs

```diff
@@ -0,0 +1,257 @@
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Collections;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.VisualStudio.LanguageServer.Protocol;
+using Microsoft.VisualStudio.Shell.Interop;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Threading;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.DocumentOutline
+{
+    internal partial class DocumentOutlineControl
+    {
+        private IWpfTextView? GetLastActiveIWpfTextView()
+        {
+            ThreadingContext.ThrowIfNotOnUIThread();
+
+            if (ErrorHandler.Failed(CodeWindow.GetLastActiveView(out var textView)))
+                return null;
+
+            return EditorAdaptersFactoryService.GetWpfTextView(textView);
+        }
+
+        /// <summary>
+        /// Starts a new task to compute the UI model.
+        /// </summary>
+        private void StartComputeUIModelTask()
+        {
+            // 'true' value is unused.  this just signals to the queue that we have work to do.
+            _computeUIModelQueue.AddWork(true);
+        }
+
+        /// <summary>
+        /// Creates the UI model.
+        /// </summary>
+        private async ValueTask<DocumentSymbolModel?> ComputeUIModelAsync(ImmutableSegmentedList<bool> unused, CancellationToken cancellationToken)
+        {
+            // Jump to the UI thread to get the currently active text view.
+            await ThreadingContext.JoinableTaskFactory.SwitchToMainThreadAsync(cancellationToken);
+
+            var activeTextView = GetLastActiveIWpfTextView();
+            if (activeTextView is null)
+                return null;
+
+            var originalSnapshot = activeTextView.TextSnapshot;
+            var textBuffer = activeTextView.TextBuffer;
+
+            var filePath = GetFilePath();
+            if (filePath is null)
+                return null;
+
+            // Ensure we switch to the threadpool before calling ComputeUIModelAsync. It ensures
+            // that fetching and processing the document model is not done on the UI thread.
+            await TaskScheduler.Default;
+
+            var model = await ComputeUIModelAsync().ConfigureAwait(false);
+
+            if (model is not null)
+                StartUpdateUIModelTask();
+
+            return model;
+
+            async Task<DocumentSymbolModel?> ComputeUIModelAsync()
+            {
+                var response = await DocumentOutlineHelper.DocumentSymbolsRequestAsync(
+                    textBuffer, LanguageServiceBroker, filePath, cancellationToken).ConfigureAwait(false);
+
+                if (response is null)
+                    return null;
```

Comment:
nit: doc when this can be null.

@dibarbet when can this be null?

Reply:
(same question below).

Reply:
I think if there is no matching LSP server registered the client will return null here - e.g. wrong content type on the buffer, the server totally failed to start, server doesn't support the right capabilities.

For C# we might know its a bug if we get a null response here, but we don't know that in general for all languages.

Similar applies to responseBody - the C# server doesn't return null, but who knows about anyone else.